File: app/src/main/java/com/tobiasfried/rasa/viewmodel/MainViewModelFactory.java
package com.tobiasfried.rasa.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.google.firebase.firestore.FirebaseFirestore;

public class MainViewModelFactory extends ViewModelProvider.NewInstanceFactory {

	private FirebaseFirestore mDb;
	private MainViewModel mainViewModel;

	public MainViewModelFactory(FirebaseFirestore db) {
		this.mDb = db;
	}

	@NonNull
	@Override
	public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
		return (T) new MainViewModel();
	}

	public MainViewModel getInstance() {
		if (mainViewModel != null) {
			return mainViewModel;
		} else {
			mainViewModel = create(MainViewModel.class);
			return mainViewModel;
		}
	}
}

File: app/src/main/java/com/tobiasfried/rasa/MainViewModelFactory.kt
package com.tobiasfried.rasa

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.google.firebase.firestore.FirebaseFirestore

class MainViewModelFactory(private val mDb: FirebaseFirestore) : ViewModelProvider.NewInstanceFactory() {
    private var mainViewModel: MainViewModel? = null

    val instance: MainViewModel
        get() {
            if (mainViewModel != null) {
                return mainViewModel
            } else {
                mainViewModel = create(MainViewModel::class.java)
                return mainViewModel
            }
        }

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return MainViewModel() as T
    }
}